server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java

```diff
@@ -315,4 +319,35 @@ private ShardRoutings(ShardId ignored) {
             this(new ArrayList<>(), new ArrayList<>());
         }
     }
+
+    private static Set<ShardRouting> getIgnoredShardsWithDiscardedAllocationStatus(List<ShardRouting> ignoredShards) {
+        return ignoredShards.stream()
+            .flatMap(shardRouting -> Stream.of(shardRouting, discardAllocationStatus(shardRouting)))
```

Comment:
I think we do not need to keep original shard routing in this set if we are also discarding status before `contains` check.